p2p/net/swarm/swarm_metrics.go

```diff
@@ -216,18 +216,30 @@ func (m *metricsTracer) CompletedHandshake(t time.Duration, cs network.Connectio
 	connHandshakeLatency.WithLabelValues(*tags...).Observe(t.Seconds())
 }
 
-func (m *metricsTracer) FailedDialing(addr ma.Multiaddr, err error) {
+func (m *metricsTracer) FailedDialing(addr ma.Multiaddr, dialErr error, cause error) {
 	transport := metricshelper.GetTransport(addr)
 	e := "other"
-	if errors.Is(err, context.Canceled) {
-		e = "canceled"
-	} else if errors.Is(err, context.DeadlineExceeded) {
+	// dial deadline exceeded or the the parent contexts deadline exceeded
+	if errors.Is(dialErr, context.DeadlineExceeded) || errors.Is(cause, context.DeadlineExceeded) {
 		e = "deadline"
+	} else if errors.Is(dialErr, context.Canceled) {
+		// dial was cancelled.
+
+		if errors.Is(cause, errParentContextCanceled) {
+			// parent was canceled
+			e = "canceled"
+		} else if errors.Is(cause, context.Canceled) {
```

Comment:
This looks brittle. Should we have a separate cause for this?

Reply:
Makes sense. That would be cleaner.

Reply:
I removed `errParentContextCanceled` and added `errConcurrentDialSuccessful`.

Now if the cause is `context.Canceled` it necessarily means the application cancelled the context and if the cause is `errConcurrentDialSuccessful` it means a concurrent dial succeeded.